extract/coordinate.py

```diff
@@ -0,0 +1,42 @@
+class Coordinate(object):
+    """Represents x/y coordinates"""
+
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+
+
+def upper_left_most_coordinate(coordinates):
```

Comment:
This could be much simpler with a list comprehension:

````python
>>> class Coord(object):
...     def __init__(self, x, y):
...         self.x = x
...         self.y = y
````
````python
>>> coords = [Coord(1,3), Coord(2,7), Coord(3,1)]
>>> max_x = max(coord.x for coord in coords)
>>> print(max_x)
3
>>> print(max(coord.y for coord in coords))
7
````

The "list comprehension" is this bit: `coord.x for coord in coords`. Under the hood, it's iterating over `coords` (using the iterator protocal you read about earlier); each element is being called `coord` in turn, `coord.x` is being called and the results are used to generate a list. Then that list is passed to `max()`

List comprehensions can be a little controversial in python: there are times (like this, I think) where they increase code readability by making it really compact and really obvious what you're doing (it's obvious you're taking the `max` of `coord.x` for all the `coord` in `coords`), but there are times when using a list comprehension can make it harder to figure out what's going on.

Some people will avoid them altogether, but I like to use them at times like this when they make the code much much simpler.

Reply:
Really you could boil this down to:

````python
def upper_left(coordinates):
    return max(coord.x for coord in coordinates), max(coord.y for coord in coordinates)
````

but I think that's gone beyond "readable" into "wtf just happened".

Also you don't actually want the max of both, you want the max of one and the min of the other, I think. But I cbf thinking about it that deep :p